runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java

```diff
@@ -159,91 +160,100 @@ private static <K, V> Partitioner getPartitioner(
    * @param <K> type of key iterator emits
    * @param <V> type of value iterator emits
    */
-  static class GroupByKeyIterator<K, V, W extends BoundedWindow>
+  abstract static class GroupByKeyIterator<K, V>
       implements Iterator<WindowedValue<KV<K, Iterable<V>>>> {
 
     private final PeekingIterator<Tuple2<ByteArray, byte[]>> inner;
-    private final Coder<K> keyCoder;
-    private final WindowingStrategy<?, W> windowingStrategy;
-    private final FullWindowedValueCoder<KV<K, V>> windowedValueCoder;
+    final Coder<K> keyCoder;
+    private ByteArray previousKey = null;
 
-    private boolean hasNext = true;
-    private ByteArray currentKey = null;
-
-    GroupByKeyIterator(
-        Iterator<Tuple2<ByteArray, byte[]>> inner,
-        Coder<K> keyCoder,
-        WindowingStrategy<?, W> windowingStrategy,
-        WindowedValue.FullWindowedValueCoder<KV<K, V>> windowedValueCoder)
-        throws Coder.NonDeterministicException {
+    GroupByKeyIterator(Iterator<Tuple2<ByteArray, byte[]>> inner, Coder<K> keyCoder) {
 
       this.inner = Iterators.peekingIterator(inner);
       this.keyCoder = keyCoder;
-      this.windowingStrategy = windowingStrategy;
-      this.windowedValueCoder = windowedValueCoder;
     }
 
     @Override
     public boolean hasNext() {
-      return hasNext;
+      return inner.hasNext();
     }
 
     @Override
     public WindowedValue<KV<K, Iterable<V>>> next() {
-      while (inner.hasNext()) {
-        final ByteArray nextKey = inner.peek()._1;
-        if (nextKey.equals(currentKey)) {
-          // we still did not see all values for a given key
+      while (hasNext()) {
+
+        final ByteArray currentKey = inner.peek()._1;
+
+        if (currentKey.equals(previousKey)) {
+          // inner iterator did not consume all values for a given key, we need to skip ahead until
+          // we find value for the next key
           inner.next();
           continue;
         }
-        currentKey = nextKey;
+        previousKey = currentKey;
+
         final WindowedValue<KV<K, V>> decodedItem = decodeItem(inner.peek());
         return decodedItem.withValue(
-            KV.of(decodedItem.getValue().getKey(), new ValueIterator(inner, currentKey)));
+            KV.of(
+                decodedItem.getValue().getKey(),
+                new Iterable<V>() {
+                  boolean consumed = false;
+
+                  @Override
+                  public Iterator<V> iterator() {
+                    if (consumed) {
+                      throw new IllegalStateException(
+                          "ValueIterator can't be iterated more than once otherwise there could be data lost");
+                    }
+                    consumed = true;
+                    return new AbstractIterator<V>() {
+
+                      @Override
+                      public V computeNext() {
+                        if (inner.hasNext() && inner.peek()._1.equals(currentKey)) {
```

Comment:
This implies that Key must use deterministic Coder, otherwise this might not work. Is there any check that the Coder is deterministic before resorting to this optimization?